core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -383,11 +385,44 @@ class TransactionCoordinator(brokerId: Int,
 
   def partitionFor(transactionalId: String): Int = txnManager.partitionFor(transactionalId)
 
+  private def expireTransactions(): Unit = {
+
+    txnManager.transactionsToExpire().foreach{ idAndMetadata =>
+      idAndMetadata.metadata synchronized {
+        if (!txnManager.isCoordinatorLoadingInProgress(idAndMetadata.transactionalId)
+          && idAndMetadata.metadata.pendingState.isEmpty) {
+          idAndMetadata.metadata.producerEpoch = (idAndMetadata.metadata.producerEpoch + 1).toShort
+          idAndMetadata.metadata.prepareTransitionTo(Ongoing)
```

Comment:
Sorry if this was raised before, but why don't we bump the epoch and transition to PREPARE_ABORT in the same write? If there's an edge case we're protecting by doing the epoch bump first, it might be useful to add a comment to document it.